xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h

```diff
@@ -294,6 +294,94 @@ class CSettingValue
 namespace kodi {
 namespace addon {
 
+/*
+ * Internally used helper class to manage processing of a "C" structure in "CPP"
+ * class.
+ *
+ * At constant, the "C" structure is copied, otherwise the given pointer is
+ * superseded and is changeable.
+ *
+ * -----------------------------------------------------------------------------
+ *
+ * Example:
+ *
+ * ~~~~~~~~~~~~~{.cpp}
+ * extern "C" typedef struct C_SAMPLE_DATA
+ * {
+ *   unsigned int iUniqueId;
+ * } C_SAMPLE_DATA;
+ *
+ * class CPPSampleData : public CStructHdl<CPPSampleData, C_SAMPLE_DATA>
+ * {
+ * public:
+ *   CPPSampleData() = default;
+ *   CPPSampleData(const CPPSampleData& sample) : CStructHdl(sample) { }
+ *   CPPSampleData(const C_SAMPLE_DATA* sample) : CStructHdl(sample) { }
+ *   CPPSampleData(C_SAMPLE_DATA* sample) : CStructHdl(sample) { }
+ *
+ *   void SetUniqueId(unsigned int uniqueId) { m_cStructure->iUniqueId = uniqueId; }
+ *   unsigned int GetUniqueId() const { return m_cStructure->iUniqueId; }
+ * };
+ *
+ * ~~~~~~~~~~~~~
+ *
+ * The class automatically works with a "C" = "CPP" or "CPP" = "C".
+ */
+template <class CPP_CLASS, typename C_STRUCT>
+class CStructHdl
+{
+public:
+  CStructHdl()
+    : m_owner(true),
+      m_cStructure(new C_STRUCT)
+  {
+  }
+
+  CStructHdl(const CPP_CLASS& cppClass)
+    : m_owner(true),
+      m_cStructure(new C_STRUCT(*cppClass.m_cStructure))
+  {
+  }
+
+  CStructHdl(const C_STRUCT* cStructure)
+    : m_owner(true),
+      m_cStructure(new C_STRUCT({*cStructure}))
+  {
+  }
+
+  CStructHdl(C_STRUCT* cStructure)
+    : m_owner(false),
+      m_cStructure(cStructure)
+  {
+  }
+
+  const CPP_CLASS& operator =(const CPP_CLASS& right)
+  {
+    m_owner = true;
+    m_cStructure = new C_STRUCT(*right.m_cStructure);
```

Comment:
```c++
if ((*right.m_cStructure)
  m_cStructure = new C_STRUCT(*right.m_cStructure);
```

Reply:
Having it with the "if" would rather bring in hidden errors.

m_cStructure would be NULL and would be called later by the child class.

Maybe for this case:
```cpp
  const CPP_CLASS& operator =(const CPP_CLASS& right)
  {
    m_owner = true;
    if (*right.m_cStructure)
      m_cStructure = new C_STRUCT(*right.m_cStructure);
    else
    {
      kodi::Log(ADDON_LOG_FATAL, "CStructHdl::%s: called with nullptr!", __func__);
      m_cStructure = new C_STRUCT;
    }
    return *this;
  }
```
But then also rather hidden and could bring confusion.

Thinking it would be safest to do so, and you would discover something in development:
```cpp
  const CPP_CLASS& operator =(const CPP_CLASS& right)
  {
    assert(*right.m_cStructure)
    m_owner = true;
    m_cStructure = new C_STRUCT(*right.m_cStructure);
    return *this;
  }
```

Reply:
Okay. If it is an error that `right` has no `m_cStructure`, then an `assert` is just fine.